Cookbook/Proposals/ComposableReducers.md

```diff
@@ -0,0 +1,181 @@
+# Composable Reducers using Lenses and Prisms
+
+* Author: Yasuhiro Inami
+* Review Manager: David Rodrigues
+
+## Introduction
+
+This proposal adds functional `Lens` and `Prism` ideas to break down `reducer` / `state` / `event` (`action`) in [ReactiveFeedback](https://github.com/Babylonpartners/ReactiveFeedback) so that we can create a large application from small component compositions.
+
+## Motivation
+
+In Babylon.app, we are using [ReactiveFeedback](https://github.com/Babylonpartners/ReactiveFeedback) to control states, side-effects, and feedback loops to define a particular screen (view controller) behavior, owned by `ViewModel`.
+
+However, this `ViewModel` can easily become too complex as `reducer: State -> Event -> State` grows large.
+
+Unfortunately, splitting into multiple `ViewModel`s is not a clever solution, as managing multiple `ReactiveFeedback`s that interact with each other tends to be hard to control.
+
+To curcumvent this problem, we instead **split `reducer`s and also `state`s and `event`s**, and then combine them using `Lens` and `Prism`.
+
+The basic idea can be found in @mbrandonw ’s talk:
+
+[Brandon Williams \- Composable Reducers & Effects Systems \- YouTube](https://www.youtube.com/watch?v=QOIigosUNGU)
+
+## Proposed solution
+
+### Functional `Lens` & `Prism`
+
+```swift
+/// e.g. A = whole struct, B = partial struct
+struct Lens<A, B> {
+    let get: (A) -> B
+    let set: (A, B) -> A
+}
+
+/// e.g. A = one of enum case, B = its associated value
+struct Prism<A, B> {
+    let preview: (A) -> B? // simplified from `A -> Either<A, B>`, a dual of `set`
+    let review: (B) -> A   // a dual of `get`
+}
+```
+
+- `Lens` is a pair of getter and setter (similar to `WritableKeyPath<A, B>`, but more composable)
+- `Prism` is a dual of `Lens`, reversing its arrows
+
+While `Lens` is useful for traversing `struct` members, `Prism` is useful for traversing `enum` cases.
+Because in ReactiveFeedback, `State` is normally defined as `struct` and `Event` is `enum`, we need both features to be able to transform `reducer` into arbitrary structure.
+
+### `Reducer`
+
+```swift
+struct Reducer<Action, State>: Monoid {
```

Comment:
Are you proposing to  introduce `Monoid` into our codebase? That would probably require a proposal of its own...

Reply:
Removed `Monoid` in 5390a38 .

Reply:
And we will have to discuss what is better, when it is modelled as a protocol or as a generic struct 😉